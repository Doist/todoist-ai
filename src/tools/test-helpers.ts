import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
