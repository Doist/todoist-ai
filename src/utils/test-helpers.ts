import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
