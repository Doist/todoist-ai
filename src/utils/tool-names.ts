/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
