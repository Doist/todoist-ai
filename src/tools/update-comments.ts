import type { Comment } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_COMMENTS, DELETE_OBJECT } = ToolNames

const ArgsSchema = {
    id: z.string().min(1).describe('The ID of the comment to update.'),
    content: z.string().min(1).describe('The new content for the comment.'),
}

const updateComments = {
    name: ToolNames.UPDATE_COMMENTS,
    description: 'Update the content of an existing comment.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const comment = await client.updateComment(args.id, { content: args.content })

        const textContent = generateTextContent({ comment })

        return getToolOutput({
            textContent,
            structuredContent: {
                comment,
                operation: 'updated',
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({ comment }: { comment: Comment }): string {
    const hasAttachment = comment.fileAttachment !== null
    const attachmentInfo = hasAttachment
        ? ` • Attachment: ${comment.fileAttachment?.fileName || 'file'}`
        : ''

    const summary = `Updated comment: ${comment.content}${attachmentInfo} • id=${comment.id}`

    // Context-aware next steps
    const nextSteps: string[] = []

    // Suggest follow-up actions based on comment type
    if (comment.taskId) {
        nextSteps.push(
            `Use ${FIND_COMMENTS} with taskId=${comment.taskId} to see all task comments`,
        )
    } else if (comment.projectId) {
        nextSteps.push(
            `Use ${FIND_COMMENTS} with projectId=${comment.projectId} to see all project comments`,
        )
    }

    nextSteps.push(`Use ${DELETE_OBJECT} with type=comment id=${comment.id} to remove comment`)

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { updateComments }
