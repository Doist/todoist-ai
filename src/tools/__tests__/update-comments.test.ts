import type { Comment, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { extractStructuredContent, extractTextContent } from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { updateComments } from '../update-comments.js'

// Mock the Todoist API
const mockTodoistApi = {
    updateComment: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { UPDATE_COMMENTS } = ToolNames

const createMockComment = (overrides: Partial<Comment> = {}): Comment => ({
    id: '12345',
    content: 'Updated comment content',
    postedAt: '2024-01-01T12:00:00Z',
    postedUid: 'user123',
    taskId: 'task123',
    projectId: undefined,
    fileAttachment: null,
    uidsToNotify: null,
    reactions: null,
    isDeleted: false,
    ...overrides,
})

describe(`${UPDATE_COMMENTS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    it('should update comment content', async () => {
        const mockComment = createMockComment({
            id: '98765',
            content: 'Updated content here',
            taskId: 'task456',
        })

        mockTodoistApi.updateComment.mockResolvedValue(mockComment)

        const result = await updateComments.execute(
            {
                id: '98765',
                content: 'Updated content here',
            },
            mockTodoistApi,
        )

        expect(mockTodoistApi.updateComment).toHaveBeenCalledWith('98765', {
            content: 'Updated content here',
        })

        // Verify result is a concise summary
        expect(extractTextContent(result)).toMatchSnapshot()

        // Verify structured content
        const structuredContent = extractStructuredContent(result)
        expect(structuredContent).toEqual(
            expect.objectContaining({
                comment: expect.objectContaining({
                    id: '98765',
                    content: 'Updated content here',
                    taskId: 'task456',
                    fileAttachment: null,
                }),
                operation: 'updated',
            }),
        )
    })

    it('should handle project comment', async () => {
        const mockComment = createMockComment({
            id: '98767',
            content: 'Updated project comment',
            taskId: undefined,
            projectId: 'project789',
        })

        mockTodoistApi.updateComment.mockResolvedValue(mockComment)

        const result = await updateComments.execute(
            {
                id: '98767',
                content: 'Updated project comment',
            },
            mockTodoistApi,
        )

        // Verify result is a concise summary
        expect(extractTextContent(result)).toMatchSnapshot()

        // Verify structured content
        const structuredContent = extractStructuredContent(result)
        expect(structuredContent).toEqual(
            expect.objectContaining({
                comment: expect.objectContaining({
                    id: '98767',
                    content: 'Updated project comment',
                    taskId: undefined,
                    projectId: 'project789',
                    fileAttachment: null,
                }),
                operation: 'updated',
            }),
        )
    })
})
