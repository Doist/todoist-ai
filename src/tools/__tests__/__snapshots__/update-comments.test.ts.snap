// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`update-comments tool should handle project comment 1`] = `
"Updated comment: Updated project comment • id=98767
Next:
- Use find-comments with projectId=project789 to see all project comments
- Use delete-object with type=comment id=98767 to remove comment"
`;

exports[`update-comments tool should update comment content 1`] = `
"Updated comment: Updated content here • id=98765
Next:
- Use find-comments with taskId=task456 to see all task comments
- Use delete-object with type=comment id=98765 to remove comment"
`;
